packages/protobuf/src/transform/index.ts

```diff
@@ -0,0 +1,933 @@
+  }
+
+  function arrayToProto(t: Model, relativeSource: Model | Operation): ProtoType {
+    const valueType = (t as Model).templateMapper!.args[0];
+
+    // Nested arrays are not supported.
+    if (isArray(valueType)) {
+      reportDiagnostic(program, {
+        code: "nested-array",
+        target: valueType,
+      });
+      return ref("<unreachable>");
+    }
+
+    return addType(valueType, relativeSource);
+  }
+
+  function getProtoScalarsMap(program: Program): Map<Type, ProtoScalar> {
+    // The type references are different object identities in different programs, so we need to cache the map per program.
+    // This really only affects tests in our current use case, but someone could be using the compiler API to compile
+    // multiple programs and it also affects that.
+    let scalarMap;
+    if (_protoScalarsMap.has(program)) {
+      scalarMap = _protoScalarsMap.get(program)!;
+    } else {
+      scalarMap = new Map<Type, ProtoScalar>(
+        (
+          [
+            [program.resolveTypeReference("TypeSpec.bytes"), scalar("bytes")],
+            [program.resolveTypeReference("TypeSpec.boolean"), scalar("bool")],
+            [program.resolveTypeReference("TypeSpec.string"), scalar("string")],
+            [program.resolveTypeReference("TypeSpec.int32"), scalar("int32")],
+            [program.resolveTypeReference("TypeSpec.int64"), scalar("int64")],
+            [program.resolveTypeReference("TypeSpec.uint32"), scalar("uint32")],
+            [program.resolveTypeReference("TypeSpec.uint64"), scalar("uint64")],
+            [program.resolveTypeReference("TypeSpec.float32"), scalar("float")],
+            [program.resolveTypeReference("TypeSpec.float64"), scalar("double")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sfixed32"), scalar("sfixed32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sfixed64"), scalar("sfixed64")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sint32"), scalar("sint32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sint64"), scalar("sint64")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.fixed32"), scalar("fixed32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.fixed64"), scalar("fixed64")],
+          ] as [[Type, unknown], ProtoScalar][]
+        ).map(([[type], scalar]) => [type, scalar])
+      );
+
+      _protoScalarsMap.set(program, scalarMap);
+    }
+    // Lazy initialize this map of known proto scalars.
+
+    return scalarMap;
+  }
+
+  function scalarToProto(t: Scalar): ProtoType {
+    const fullName = getTypeName(t);
+
+    const protoType = getProtoScalarsMap(program).get(t);
+
+    if (!protoType) {
+      reportDiagnostic(program, {
+        code: "unsupported-field-type",
+        messageId: "unknown-scalar",
+        format: {
+          name: fullName,
+        },
+        target: t,
+      });
+      return unreachable("unknown scalar");
```

Comment:
Tehcnically here you'd probably want to traverse up the scalar base until you find a scalar that you know.

```
scalar MyInt32 extends int32;
```

that should still resolve to int32